### PR TITLE
Finished Embedded Teaser Training

### PR DESCRIPTION
**Pull request by MarceloLiGonzales**
Assumptions of what valid input means:
Both motors must be on or off at the same time.
If both motors are off, velocities must be zero.
If motors are on, both velocities must have the same direction but not necessarily the same magnitude.
The direction of the motors and the velocities must match, forward should mean positive velocity and vice-versa.

You can't have both right and left turn lights on at the same time.
If headlights are off, neither low nor high headlights can be on.
If headlights are on, at least one, or both, of the modes should be on.

### CODE REVIEW

File: Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Core/Src/Training.c [hunk 1]
```diff
@@ -1,6 +1,73 @@
...
+
+        if (getbit(motor2, 0) != !getbit(motor2, 6)) // check motor 2 direction matches its velocity
+        *validity &= ~0b00000010;
+    }
+
+}
+
+void checklights(uint8_t *validity, uint8_t lights)
+{
+    if (getbit(lights, 3) && getbit(lights, 4))     //check that not both turn signals are on
```
Comment (badeniran) [inline Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Core/Src/Training.c:63]: Both signals can be on if the hazard are on

File: Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Core/Src/Training.c [hunk 2]
```diff
@@ -1,6 +1,73 @@
...
+    *validity &= ~0b00000100;
+
+    if (getbit(lights, 0)) {                        //if headlights are off, check all modes are off
+        if (getbit(lights, 1) || getbit(lights, 2))
+        *validity &= ~0b00000100;
+    } else {                                        // if headlights are on, check at least one mode is on
+        if (!(getbit(lights, 1) || getbit(lights, 2)))
```
Comment (badeniran) [inline Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Core/Src/Training.c:70]: Both modes can't be on at the same time.

File: Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Test/TrainingTest.c [hunk 3]
```diff
@@ -5,24 +5,37 @@
...
+    RUN_TEST(test_OnlyLightsInvalid);
+    RUN_TEST(test_OnlyMotorsInvalid);
 }
 
 void test_EverythingValid()
 {
-
+    uint8_t InputArray[3] = {0b00000000, 0b00000000, 0b00000001};
+    trainingTask(InputArray);
+    TEST_ASSERT_EQUAL_MESSAGE(0b00000111, validData, "All valid test failed");
```
Comment (badeniran) [inline Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Test/TrainingTest.c:18]: Need to test `outputArray` as well

**Review by badeniran — CHANGES_REQUESTED** [review]
A couple of logical changes, but looks good! Great commenting, and I like the way you've split things up into smaller functions

File: Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Test/TrainingTest.c [hunk 4]
```diff
@@ -5,24 +5,37 @@
...
 }
 
 void test_EverythingInvalid()
 {
+    uint8_t InputArray[3] = {0b01000011, 0b010000011, 0b00000011};
+    trainingTask(InputArray);
+    TEST_ASSERT_EQUAL_MESSAGE(0b00000000, validData, "All invalid test failed");
```
Comment (JessieG123) [inline Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Test/TrainingTest.c:25]: A better message would be like "data is not blah" that way you know exactly what is wrong when the test fails.

**Review by JessieG123 — CHANGES_REQUESTED** [review]

File: Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Test/TrainingTest.c (hunk 4 above)
Comment (MarceloLiGonzales) [reply Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Test/TrainingTest.c:25]: So in this case I should put "Data is not 0b00000000"?

Comment (JessieG123) [reply Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Test/TrainingTest.c]: You could say something like "validData is not 0x0" or "Should not have any valid data". You want to tell the reader what value it _should_ have if the test fails.

**Review by islamrahat735 — APPROVED** [review]

File: Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Core/Inc/Training.h [hunk 5]
```diff
@@ -5,3 +5,11 @@ uint8_t outputArray[3];
...
 
 void trainingTask(uint8_t* data); // data should be 3 elements wide (size of 3)
+
+uint8_t getbit(uint8_t data,int number);
+
+int getvelocity(uint8_t data);
+
+void checkmotors(uint8_t *validity, uint8_t motor1, uint8_t motor2);
+
+void checklights(uint8_t *validity, uint8_t lights);
```
Comment (duan-le) [inline Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Core/Inc/Training.h:15]: You should use camelCase to name your functions.

File: Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Core/Src/Training.c [hunk 6]
```diff
@@ -1,6 +1,95 @@
...
+
+    int Motor1State = getbit(motor1, 7);
+    if (Motor1State != getbit(motor2, 7)){  //check both motors are either turned on or off
+        *validity &= ~0b00000011;
+        return;
+    } 
+
+    if(Motor1State == 0){ //if motors are off, check velocity is 0
+
+        if (velocity1 != 0)
```
Comment (duan-le) [inline Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Core/Src/Training.c:54]: Make sure that your spacing and styling are consistent. For example, the spacing between the `if` and the `()`.

File: Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Core/Src/Training.c [hunk 7]
```diff
@@ -1,6 +1,95 @@
...
+    return (data & 1<<number);
+}
+
+int getVelocity(uint8_t data)
+{
+    int velocityBits  = getBit(data, 1) + getBit(data, 2) + getBit(data, 3) + getBit(data, 4) + getBit(data, 5);
+    int velocity = velocityBits >> 1;
+
+    if (getBit(data, 6))
+    velocity *=-1;
```
Comment (duan-le) [inline Recruit-Training/Advanced-Recruit-Training/Embedded-Teaser/Training/Core/Src/Training.c:36]: This should be indented. You should put this in curly braces even if it's just one line. Do the same for all other lines that this applies to. Other than that, everything else looks good!

**Review by duan-le — APPROVED** [review]
Looks good! You can close this PR now.

**Pull request closed without merging**